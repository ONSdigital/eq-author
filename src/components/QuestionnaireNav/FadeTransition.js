import React from "react";
import PropTypes from "prop-types";
import { Transition } from "react-transition-group";

const FadeTransition = ({
  children,
  component: Component,
  duration,
  ...otherProps
}) =>
  <Transition timeout={duration} {...otherProps}>
    {state =>
      <Component state={state}>
        {children}
      </Component>}
  </Transition>;

FadeTransition.propTypes = {
  children: PropTypes.node.isRequired,
  component: PropTypes.any.isRequired, // eslint-disable-line react/forbid-prop-types
  duration: PropTypes.number.isRequired
};

export default FadeTransition;
