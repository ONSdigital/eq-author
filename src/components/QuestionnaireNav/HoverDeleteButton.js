import styled from "styled-components";
import { colors } from "constants/theme";

const HoverDeleteButton = styled.button`
  color: ${colors.text};
  border: none;
  background: transparent;
  font-size: 1.25em;
  padding: 0.125em 0.625em 0.125em 0.375em;
  position: absolute;
  right: 0;
  z-index: 3;
  cursor: pointer;
  transform: translateX(50%);
  opacity: 0;
  transition: transform 0.1s ease-in, opacity 0.1s ease-in;
`;

export default HoverDeleteButton;
