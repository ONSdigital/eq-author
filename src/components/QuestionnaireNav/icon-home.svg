<svg width="20" height="20" viewBox="0 0 24 24">
    <path
            fill="#fff"
            d="M10,20V14H14V20H19V12H22L12,3L2,12H5V20H10Z"
    />
</svg>