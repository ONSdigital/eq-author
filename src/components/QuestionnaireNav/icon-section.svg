<?xml version="1.0" encoding="UTF-8"?>
<svg width="14px" height="16px" viewBox="0 0 14 16" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 45.2 (43514) - http://www.bohemiancoding.com/sketch -->
    <title>Group 5</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Group-5" transform="translate(-2.000000, -1.000000)">
            <g id="ic_insert_drive_file_black_24px">
                <path d="M4,1.33333333 C3.26666667,1.33333333 2.67333333,1.93333333 2.67333333,2.66666667 L2.66666667,13.3333333 C2.66666667,14.0666667 3.26,14.6666667 3.99333333,14.6666667 L12,14.6666667 C12.7333333,14.6666667 13.3333333,14.0666667 13.3333333,13.3333333 L13.3333333,5.33333333 L9.33333333,1.33333333 L4,1.33333333 Z" id="Shape" fill-opacity="0.54" fill="#4A4A4A" fill-rule="nonzero"></path>
                <path d="M15.3333333,6 L7.33333333,6 C6.6,6 6,6.6 6,7.33333333 L6,16.6666667 L7.33333333,16.6666667 L7.33333333,7.33333333 L15.3333333,7.33333333 L15.3333333,6 Z" id="Path" fill-opacity="0.54" fill="#4A4A4A" transform="translate(10.666667, 11.333333) rotate(-180.000000) translate(-10.666667, -11.333333) "></path>
                <polygon id="Shape" points="0 0 16 0 16 16 0 16"></polygon>
            </g>
        </g>
    </g>
</svg>