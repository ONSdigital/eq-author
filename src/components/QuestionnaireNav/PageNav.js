import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { getLink } from "utils/UrlUtils";
import Tooltip from "components/Tooltip";
import CustomPropTypes from "custom-prop-types";
import { TransitionGroup, CSSTransition } from "react-transition-group";
import { NavLink } from "react-router-dom";
import { colors } from "constants/theme";
import getTextFromHTML from "utils/getTextFromHTML";
import HoverDeleteButton from "./HoverDeleteButton";

const duration = 300;

export const StyledPageItem = styled.li`
  padding: 0;
  margin: 0;
  font-weight: 500;
  display: flex;
  align-items: center;
  z-index: ${props => props.index};
  position: relative;

  &.page-enter,
  &.page-exit.page-exit-active {
    opacity: 0;
    height: 0;
    transform: translateX(-20px);
  }

  &.page-exit,
  &.page-enter.page-enter-active {
    opacity: 1;
    height: 2em;
    transform: translateX(0);
  }

  &.page-enter-active {
    transition: height ${duration / 2}ms ease-out,
      opacity ${duration / 2}ms ease-out ${duration / 2}ms,
      transform ${duration / 2}ms ease-out ${duration / 2}ms;
  }

  &.page-exit-active {
    transition: opacity ${duration / 2}ms ease-out,
      transform ${duration / 2}ms ease-out,
      height ${duration / 2}ms ease-out ${duration / 2}ms;
  }
`;

const Link = styled(NavLink)`
  text-decoration: none;
  color: ${colors.text};
  font-size: 0.75em;
  padding: 0.7em 2.5em 0.7em 0.9em;
  display: block;
  flex: 1 1 auto;
  height: 100%;
  position: relative;
  overflow: hidden;
  transition: opacity 100ms ease-out;

  &::before {
    opacity: 0;
    content: "";
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 0;
    transition: all 200ms ease-out;
    background: ${colors.borders};
    z-index: 1;
  }

  /* stylelint-disable */
  ${StyledPageItem}:hover & {
    &::before {
      opacity: 0.5;
      width: 100%;
    }
  }
  /* stylelint-enable */

  &.selected {
    &::before {
      opacity: 1 !important;
      width: 100%;
    }
  }

  &[aria-disabled="true"] {
    pointer-events: none;
    opacity: 0.5;
  }
`;

export const LinkText = styled.span`
  display: inline-block;
  vertical-align: middle;
  width: 100%;
  position: relative;
  z-index: 2;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  opacity: ${({ fade }) => (fade ? 0.5 : 1)};
`;

const NavList = styled.ol`
  padding: 0;
  margin: 0;
  list-style: none;
`;

export const DeleteButton = styled(HoverDeleteButton)`
  top: 0.1em;
  /* stylelint-disable */
  ${StyledPageItem}:hover &,
  ${Link}:focus + &,
  &:focus {
    opacity: 1;
    transform: translateX(0);
  }
  /* stylelint-enable */
`;

export class PageNavItem extends React.Component {
  state = {
    isDeleting: false
  };

  static propTypes = {
    sectionId: PropTypes.string.isRequired,
    questionnaireId: PropTypes.string.isRequired,
    pageId: PropTypes.string.isRequired,
    pageNumber: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    onDelete: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired
  };

  handleDelete = () => {
    const { sectionId, pageId, onDelete } = this.props;

    this.setState({ isDeleting: true });

    onDelete(sectionId, pageId).catch(() => {
      this.setState({ isDeleting: false });
    });
  };

  render() {
    const {
      sectionId,
      questionnaireId,
      pageId,
      title,
      pageNumber,
      index
    } = this.props;

    return (
      <StyledPageItem index={index}>
        <Link
          to={getLink(questionnaireId, sectionId, pageId)}
          aria-disabled={parseInt(pageId, 10) < 0}
          activeClassName="selected"
        >
          <LinkText fade={this.state.isDeleting}>
            {pageNumber} {getTextFromHTML(title) || "Page Title"}
          </LinkText>
        </Link>
        <Tooltip content="Delete page">
          <DeleteButton
            type="button"
            aria-label="Delete page"
            onClick={this.handleDelete}
          >
            Ã—
          </DeleteButton>
        </Tooltip>
      </StyledPageItem>
    );
  }
}

const PageNav = ({ section, questionnaire, onDelete }) => (
  <TransitionGroup component={NavList}>
    {section.pages.map((page, i, pages) => {
      const pageNumber = `${section.number}.${i + 1}`;
      return (
        <CSSTransition key={page.id} timeout={duration} classNames="page">
          <PageNavItem
            title={page.title}
            pageNumber={pageNumber}
            pageId={page.id}
            sectionId={section.id}
            questionnaireId={questionnaire.id}
            onDelete={onDelete}
            index={pages.length - i}
          />
        </CSSTransition>
      );
    })}
  </TransitionGroup>
);

PageNav.propTypes = {
  questionnaire: CustomPropTypes.questionnaire.isRequired,
  section: CustomPropTypes.section.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default PageNav;
