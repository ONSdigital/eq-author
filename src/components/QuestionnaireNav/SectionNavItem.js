import React from "react";
import styled from "styled-components";

import { colors } from "constants/theme";
import HoverDeleteButton from "components/QuestionnaireNav/HoverDeleteButton";
import PageNav from "components/QuestionnaireNav/PageNav";
import SectionTitle from "components/QuestionnaireNav/SectionTitle";
import Tooltip from "components/Tooltip";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";

export const AddPageBtn = styled.button`
  appearance: none;
  cursor: pointer;
  background: none;
  border: none;
  padding: 1em;
  color: ${colors.text};
  &:hover {
    color: black;
  }
`;

export const SectionDeleteButton = styled(HoverDeleteButton)`
  top: 0.5em;

  .section-title-wrapper:hover &,
  .section-title-wrapper:focus + &,
  &:focus {
    opacity: 1;
    transform: translateX(0);
  }
`;

const halfDuration = props => props.duration / 2;
const duration = props => props.duration;

const StyledSectionNavItem = styled.li`
  margin: 0;
  padding: 0.5em 0;
  position: relative;

  &:not(:last-child) {
    border-bottom: 1px solid #c3c3c3;
  }

  &.section-enter {
    opacity: 0;
    transform: translateX(-20px);
  }

  &.section-enter-active {
    opacity: 1;
    transform: translateX(0);
    transition: opacity ${duration}ms ease-out, transform ${duration}ms ease-out;
  }

  &.section-exit {
    opacity: 1;
    transform: translateX(0);
  }

  &.section-exit-active {
    opacity: 0;
    transform: translateX(-20px);
    height: 0 !important;
    transition: opacity ${halfDuration}ms ease-out,
      transform ${halfDuration}ms ease-out,
      height ${halfDuration}ms ease-in ${halfDuration}ms;
  }
`;

StyledSectionNavItem.propTypes = {
  duration: PropTypes.number.isRequired
};

class SectionNavItem extends React.Component {
  static propTypes = {
    questionnaire: CustomPropTypes.questionnaire,
    onAddPage: PropTypes.func.isRequired,
    onDeleteSection: PropTypes.func.isRequired,
    onDeletePage: PropTypes.func.isRequired,
    section: CustomPropTypes.section.isRequired,
    duration: PropTypes.number.isRequired,
    saveSectionItemRef: PropTypes.func.isRequired
  };

  state = {
    style: {}
  };

  handleAddPage = () => {
    this.props.onAddPage(this.props.section.id);
  };

  handleDeleteSection = () => {
    const { height } = this.elem.getBoundingClientRect();
    const style = { height };

    this.setState({ style }, () =>
      this.props.onDeleteSection(this.props.section.id)
    );
  };

  saveRef = elem => {
    const { section, saveSectionItemRef } = this.props;
    this.elem = elem;
    saveSectionItemRef(section.id, elem);
  };

  render() {
    const { questionnaire, section, onDeletePage, duration } = this.props;

    return (
      <StyledSectionNavItem
        innerRef={this.saveRef}
        style={this.state.style}
        duration={duration}
      >
        <div className="section-title-wrapper">
          <SectionTitle questionnaire={questionnaire} section={section} />
          <Tooltip content="Delete section" offset={{ right: -5 }}>
            <SectionDeleteButton
              type="button"
              aria-label="Delete section"
              onClick={this.handleDeleteSection}
            >
              &times;
            </SectionDeleteButton>
          </Tooltip>
        </div>
        <PageNav
          section={section}
          questionnaire={questionnaire}
          onDelete={onDeletePage}
        />
        <AddPageBtn onClick={this.handleAddPage} id="btn-add-page">
          + Add page
        </AddPageBtn>
      </StyledSectionNavItem>
    );
  }
}

export default SectionNavItem;
