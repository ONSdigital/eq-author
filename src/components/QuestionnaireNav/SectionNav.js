import React, { Component } from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";

import styled from "styled-components";
import { colors } from "constants/theme";

import sectionIcon from "./icon-section.svg";

import PageNav from "components/QuestionnaireNav/PageNav";
import SectionNavItem from "components/QuestionnaireNav/SectionNavItem";
import { TransitionGroup } from "react-transition-group";

import { NavLink } from "react-router-dom";

import { first, delay } from "lodash";
import { getLink } from "utils/UrlUtils";

const Link = styled(NavLink)`
  text-decoration: none;
  &:link,
  &:visited {
    color: ${colors.text};
  }
`;

const SectionTitle = styled.h3`
  padding: 0.5em 0;
  font-size: 0.75em;
  margin: 0;
  font-weight: 900;
  display: flex;
  &:before {
    content: url(${sectionIcon});
    margin-right: 0.5em;
  }
`;

export const AddPageBtn = styled.button`
  appearance: none;
  cursor: pointer;
  background: none;
  border: none;
  padding: 1em;
  color: ${colors.text};
  &:hover {
    color: black;
  }
`;

const NavList = styled.ol`
  margin: 0;
  padding: 0;
  list-style: none;
`;

export const LinkedSectionTitle = ({ questionnaire, section }) => {
  const sectionTitle = (
    <SectionTitle>{section.title || "Section Title"}</SectionTitle>
  );
  if (section.pages.length > 0) {
    const firstPage = first(section.pages);
    return (
      <Link
        to={getLink(questionnaire.id, section.id, firstPage.id)}
        aria-disabled={parseInt(firstPage.id, 10) < 0}
        activeClassName="selected"
      >
        {sectionTitle}
      </Link>
    );
  } else {
    return sectionTitle;
  }
};

LinkedSectionTitle.propTypes = {
  questionnaire: CustomPropTypes.questionnaire.isRequired,
  section: CustomPropTypes.section.isRequired
};

class SectionNav extends Component {
  sectionItems = [];

  static propTypes = {
    questionnaire: CustomPropTypes.questionnaire,
    onAddPage: PropTypes.func.isRequired,
    onDeletePage: PropTypes.func.isRequired
  };

  scrollSectionIntoView = id => {
    delay(
      node => {
        node.scrollIntoView({ behavior: "smooth" });
      },
      200,
      this.sectionItems[id]
    );
  };

  saveSectionItemRef = (id, node) => {
    this.sectionItems[id] = node;
  };

  render() {
    const { questionnaire, onAddPage, onDeletePage } = this.props;
    return (
      <TransitionGroup component={NavList}>
        {questionnaire.sections
          .map((section, i) => ({
            ...section,
            number: `${i + 1}.`
          }))
          .map((section, sectionNum) => (
            <SectionNavItem
              key={section.number}
              section={section}
              saveSectionItemRef={this.saveSectionItemRef}
            >
              <LinkedSectionTitle
                questionnaire={questionnaire}
                section={section}
              />
              <PageNav
                section={section}
                questionnaire={questionnaire}
                onDelete={onDeletePage}
              />
              <AddPageBtn
                onClick={function() {
                  onAddPage(section.id);
                }}
                id="btn-add-page"
              >
                + Add page
              </AddPageBtn>
            </SectionNavItem>
          ))}
      </TransitionGroup>
    );
  }
}

export default SectionNav;
