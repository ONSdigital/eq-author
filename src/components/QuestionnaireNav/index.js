import React from "react";

import styled from "styled-components";

import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";

import { colors } from "constants/theme";

import SectionNav from "components/QuestionnaireNav/SectionNav";

import plusIcon from "./icon-plus.svg";

const Container = styled.div`
  padding: 1em;
  margin: 0;
`;

const Title = styled.h2`
  font-size: 0.6em;
  text-transform: uppercase;
  font-weight: 900;
  margin: 0;
  line-height: 1.5;
`;

const AddSection = styled.div`
  padding: 1em 0.5em;
  border-bottom: 1px solid ${colors.lightGrey};
`;

export const AddSectionBtn = styled.button`
  appearance: none;
  cursor: pointer;
  background: none;
  border: none;
  padding: 0;
  color: ${colors.text};
  font-weight: 600;
  font-size: 0.75em;
  &::before {
    vertical-align: middle;
    display: inline-block;
    content: url(${plusIcon});
    margin-right: 0.6em;
  }
  &:hover {
    color: black;
  }
`;

const QuestionnaireNav = ({
  questionnaire,
  onAddPage,
  onAddSection,
  onDeletePage
}) =>
  <Container>
    <Title>Questionnaire structure</Title>
    <SectionNav
      transitionDuration={200}
      questionnaire={questionnaire}
      onAddPage={onAddPage}
      onDeletePage={onDeletePage}
    />
    <AddSection>
      <AddSectionBtn
        primary
        onClick={function() {
          onAddSection(questionnaire.id);
        }}
      >
        Create new section
      </AddSectionBtn>
    </AddSection>
  </Container>;

QuestionnaireNav.propTypes = {
  questionnaire: CustomPropTypes.questionnaire.isRequired,
  onAddPage: PropTypes.func.isRequired,
  onDeletePage: PropTypes.func.isRequired,
  onAddSection: PropTypes.func.isRequired
};
export default QuestionnaireNav;
