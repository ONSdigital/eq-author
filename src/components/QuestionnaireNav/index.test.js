import React from "react";
import { shallow } from "enzyme";
import QuestionnaireNav, { AddSectionBtn } from "components/QuestionnaireNav";

describe("QuestionnaireNav", () => {
  let wrapper;
  const questionnaire = { id: "1" };
  const handleAddSection = jest.fn();
  const handleAddPage = jest.fn();
  const handleDeletePage = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <QuestionnaireNav
        questionnaire={questionnaire}
        onAddPage={handleAddPage}
        onAddSection={handleAddSection}
        onDeletePage={handleDeletePage}
      />
    );
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle clicks on `create a new section`", () => {
    wrapper.find(AddSectionBtn).simulate("click");
    expect(handleAddSection).toHaveBeenCalledWith(questionnaire.id);
  });
});
