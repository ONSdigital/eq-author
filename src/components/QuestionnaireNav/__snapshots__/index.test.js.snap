// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionnaireNav should render 1`] = `
<QuestionnaireNav__Container
  id="questionnaire-nav"
>
  <QuestionnaireNav__Title>
    Questionnaire structure
  </QuestionnaireNav__Title>
  <SectionNav
    onAddPage={[Function]}
    onDeletePage={[Function]}
    onDeleteSection={[Function]}
    questionnaire={
      Object {
        "id": "1",
        "sections": Array [
          Object {
            "id": "3",
            "pages": Array [
              Object {
                "id": "2",
                "title": "Page",
              },
            ],
            "title": "Section",
          },
        ],
        "title": "Questionnaire",
      }
    }
    transitionDuration={200}
  />
  <QuestionnaireNav__AddSection>
    <QuestionnaireNav__AddSectionBtn
      onClick={[Function]}
      primary={true}
    >
      Create new section
    </QuestionnaireNav__AddSectionBtn>
  </QuestionnaireNav__AddSection>
</QuestionnaireNav__Container>
`;
